### ISSUE TITLE
CUDA_ERROR_OUT_OF_MEMORY on tilenight stdstar step

### ISSUE THREAD

**Issue by sbailey** [desihub/desispec#1897]
desi_proc_tilenight runs out of GPU memory during the desi_fit_stdstars step.  This especially occurs when it also ran GPU extractions earlier in the same script invocation, indicating the cupy isn't releasing memory from the extraction step.   If you re-run desi_proc_tilenight a second time so that it skips over the already completed extractions, it succeeds on some petals but still runs out of memory on others (See note 2 below).

Using only 40 MPI ranks instead of 64 on a second desi_proc_tilenight invocation avoids the problem, but it still runs out of memory using 40 ranks if starting from the beginning with extractions.

@dmargala could you take a look at this and advise (or even fix?).  @marcelo-alvarez @craigwarner-ufastro heads up too.

### Steps to reproduce

```
source /global/cfs/cdirs/desi/software/desi_environment.sh main

# cd somewhere, then...
export DESI_SPECTRO_REDUX=$(pwd)
export SPECPROD=gpu_mem
copyprod $DESI_ROOT/spectro/redux/daily $SPECPROD -n 20211220 -t 22111

cd $SPECPROD
rm exposures/20211220/00114851/*.*

salloc -N 1 -C gpu -A desi_g --gpus-per-node=4 -t 01:00:00 -q interactive

# runs out of memory, hangs; a second invocation will get further but still hang
time srun -N 1 -n 64 -c 2 --cpu-bind=cores desi_mps_wrapper \
    desi_proc_tilenight -n 20211220 -t 22111 --gpuspecter --gpuextract --mpi --mpistdstars &> tilenight-1.log
```

### Note 1

See https://docs.cupy.dev/en/stable/user_guide/memory.html for discussion about cupy memory management.  `cupy.get_default_memory_pool().free_all_blocks()` can force it to release memory, but it's unclear to me whether every MPI rank should do that, or one per GPU, or something else.  Something along this line would probably be helpful between major steps of the pipeline, e.g. to ensure that all extraction GPU memory is freed before moving on.

### Note 2

`desi_proc_tilenight` -> `desi_proc_joint_fit` splits the MPI communicator into one subcommunicator per petal and then calls -> `runcmd` for `desispec.scripts.stdstars.main` with that subcommunicator.  Normally `runcmd` assumes that if one rank raises an exception, they all do, and it only logs the traceback for rank 0 to avoid printing the same traceback N>>1 times.  However, in this case, a non-subrank 0 can be the one raising a CUDE_ERROR_OUT_OF_MEMORY problem and with current desispec main that doesn't get logged.

Change line 153 in py/desispec/util.py inside `runcmd` to have it print tracebacks for every rank.

### Note 3

The stdstar GPU usage is via `scripts.stdstars.main` -> `fluxcalibration.match_templates` -> `fluxcalibration.applySmoothingFilter`, which gets called a lot on both data and templates.  I haven't fully traced the structure, but I wouldn't be surprised if we were inadvertently smoothly the same template multiple times for different stars, or the same star multiple times for different templates, or both.  GPU support was also post-facto added and it is doing one spectrum at a time.  We might save on both memory and speed if we  were able to do these smoothing steps in 2D[nspec,nwave] bulk.

That could be a deeper refactor; in the short term we need something that doesn't blow memory so that we can proceed with productions on Perlmutter, which might even include turning off GPUs for stdstars.

### Note 4

Why didn't we see this before?  Perhaps we we tested on dark tiles with fewer standards, or tested GPU stdstar fitting in isolation.  Also, until PR #1883 a week ago, desi_proc_joint_fit also accidentally wasn't using GPUs on standard stars, so bulk tilenight runs weren't hitting this.

**Comment by dmargala** [comment]
I added a few lines to proc_tilenight just before the stdstar step to help debug:


```diff
diff --git a/py/desispec/scripts/proc_tilenight.py b/py/desispec/scripts/proc_tilenight.py
index ddf40880..9677f61e 100644
--- a/py/desispec/scripts/proc_tilenight.py
+++ b/py/desispec/scripts/proc_tilenight.py
@@ -180,6 +180,11 @@ def main(args=None, comm=None):
             log.info(f'prestdstar failed with {error_count_pre} errors, skipping rest of tilenight steps')
         continue_tilenight=False

+    import cupy
+    mempool = cupy.get_default_memory_pool()
+    device_id = cupy.cuda.runtime.getDevice()
+    log.info(f"{comm.rank=} {device_id=} {mempool.used_bytes()} {mempool.total_bytes()}")
+
     if continue_tilenight:
         #- run joint stdstar fit using all exp for this tile night
         stdstar_args  = common_args + mpi_args
```

here are the extra logs that come out when I run your reproducer with that patch to desispec:

```
INFO:proc_tilenight.py:186:main: comm.rank=0 device_id=0 0 540156928
INFO:proc_tilenight.py:186:main: comm.rank=1 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=2 device_id=0 466432 5992921600
INFO:proc_tilenight.py:186:main: comm.rank=3 device_id=0 466432 6033458176
INFO:proc_tilenight.py:186:main: comm.rank=4 device_id=0 466432 5988320256
INFO:proc_tilenight.py:186:main: comm.rank=5 device_id=0 466432 6043324416
INFO:proc_tilenight.py:186:main: comm.rank=6 device_id=0 466432 5978721280
INFO:proc_tilenight.py:186:main: comm.rank=7 device_id=1 466432 5875358208
INFO:proc_tilenight.py:186:main: comm.rank=8 device_id=1 466432 5990491136
INFO:proc_tilenight.py:186:main: comm.rank=9 device_id=1 466432 5873332736
INFO:proc_tilenight.py:186:main: comm.rank=10 device_id=1 466432 5846725120
INFO:proc_tilenight.py:186:main: comm.rank=11 device_id=1 466432 5864357888
INFO:proc_tilenight.py:186:main: comm.rank=12 device_id=2 466432 5783403520
INFO:proc_tilenight.py:186:main: comm.rank=13 device_id=2 466432 5810460160
INFO:proc_tilenight.py:186:main: comm.rank=14 device_id=2 466432 5934780416
INFO:proc_tilenight.py:186:main: comm.rank=15 device_id=2 466432 6000399360
INFO:proc_tilenight.py:186:main: comm.rank=16 device_id=2 466432 6033098752
INFO:proc_tilenight.py:186:main: comm.rank=17 device_id=3 466432 5981291520
INFO:proc_tilenight.py:186:main: comm.rank=18 device_id=3 466432 5896784896
INFO:proc_tilenight.py:186:main: comm.rank=19 device_id=3 466432 5980579328
INFO:proc_tilenight.py:186:main: comm.rank=20 device_id=3 466432 6045556736
INFO:proc_tilenight.py:186:main: comm.rank=21 device_id=3 466432 6028824064
INFO:proc_tilenight.py:186:main: comm.rank=22 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=23 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=24 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=25 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=26 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=27 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=28 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=29 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=30 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=31 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=32 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=33 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=34 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=35 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=36 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=37 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=38 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=39 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=40 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=41 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=42 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=43 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=44 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=45 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=46 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=47 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=48 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=49 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=50 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=51 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=52 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=53 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=54 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=55 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=56 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=57 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=58 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=59 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=60 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=61 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=62 device_id=0 0 0
INFO:proc_tilenight.py:186:main: comm.rank=63 device_id=0 0 0
```

So it looks like the MPI ranks are not spread out across GPUs for the stdstar step. The device_id mapping for rank 0-21 is leftover from the gpu extraction step (rank 0 and 1 are special IO ranks and then 2-21 are mapped across the 4 GPUs in blocks). The GPU memory pool from extraction step expanded up to just under 6 GB per rank but there is less than 0.5 MB of that used (I'm not sure what that is but it doesn't seem horrible).

Adding a bit more to free the existing mempool and force all MPI ranks in the comm to choose a device in a cyclic/round-robin fashion:

```diff
diff --git a/py/desispec/scripts/proc_tilenight.py b/py/desispec/scripts/proc_tilenight.py
index ddf40880..92db79c8 100644
--- a/py/desispec/scripts/proc_tilenight.py
+++ b/py/desispec/scripts/proc_tilenight.py
@@ -180,6 +180,15 @@ def main(args=None, comm=None):
             log.info(f'prestdstar failed with {error_count_pre} errors, skipping rest of tilenight steps')
         continue_tilenight=False

+    import cupy
+    mempool = cupy.get_default_memory_pool()
+    device_id = cupy.cuda.runtime.getDevice()
+    log.info(f"{comm.rank=} {device_id=} {mempool.used_bytes()} {mempool.total_bytes()}")
+    mempool.free_all_blocks()
+    new_device_id = comm.rank % 4
+    cupy.cuda.Device(new_device_id).use()
+    log.info(f"{comm.rank=} {new_device_id=}")
+
     if continue_tilenight:
         #- run joint stdstar fit using all exp for this tile night
         stdstar_args  = common_args + mpi_args
```

Applying that patch and rerunning the example tilenight seems to resolve the issue:

```
> rm exposures/20211220/00114851/*.*
> time srun -N 1 -n 64 -c 2 --cpu-bind=cores desi_mps_wrapper     desi_proc_tilenight -n 20211220 -t 22111 --gpuspecter --gpuextract --mpi --mpistdstars &> tilenight-2.log

real	8m47.221s
user	0m0.343s
sys	0m1.411s
> grep ERROR tilenight-2.log
>
```

**Comment by craigwarner-ufastro** [comment]
Great!  If there are any remaining issues I'd be happy to help take a 
look at them next week.  I'll just note that cupy can be slow to release 
memory - I explicitly call free_all_blocks() in my GPU version of 
calc_zchi2_batch - although only for consumer grade GPUs with lower 
memory (< 3 GB free as of the start of the method) because it does come 
at a small time cost.  The A100s have 40 or 80 GB depending on the chip 
so there are methods that will run fine on them but will crash on 
consumer grade GPUs with less memory - something we should keep in mind 
for anything that might be run on individual machines and not just on 
Perlmutter.

Cheers,

Craig

**Comment by sbailey** [comment]
@dmargala wow, that looks great.  Thanks for the quick debugging and suggested fixes.  I think that gives me the pieces to put it into the pipeline (including handling the non-GPU cases etc).

Is the extraction rank:device mapping explicitly assigned somewhere, or is that just how it was auto-assigned during that step when the GPUs first got used, and then stuck around for later steps?

**Comment by dmargala** [comment]
by default, each process will use device 0. 

the extraction mapping seen just before the stdstar step is leftover from the extraction step. it's explicitly assigned by this function of gpu_specter: https://github.com/desihub/gpu_specter/blob/77e3db442d8a7b8ef00d21437293968494c326e7/py/gpu_specter/core.py#L349

**Comment by sbailey** [comment]
@dmargala could you check your tilenight-2.log (and/or ping me with the location)?  My tests are finishing (i.e. not hanging), but have an exit code of 143, aren't actually running to completion, and include these in the logs just before abruptly ending in sky subtraction during calculate_throughput_corrections:
```
srun: error: nid002781: tasks 0,2-21: Terminated
srun: Force Terminated StepId=3588045.1
```
Curiously, that is crashing earlier than it was yesterday, apparently even before the cupy memory flushing code and device re-allocation even happens.  Did some slurm auto-kill parameter change in the last 24 hours related to one of the outages?

**Comment by sbailey** [comment]
Fixed by #1899.  Closing.
